File: plugins/plugin-auto-trader/python/elizaos_plugin_auto_trader/portfolio.py
from __future__ import annotations

from elizaos_plugin_auto_trader.types import Holding, Portfolio, Trade


class PortfolioManager:
    """Manages an in-memory portfolio: holdings and trade history."""

    def __init__(self, initial_value: float = 10_000.0) -> None:
        self._holdings: dict[str, Holding] = {}
        self._trade_history: list[Trade] = []
        self._initial_value = initial_value

    def get_portfolio(self) -> Portfolio:
        total_value = sum(h.value for h in self._holdings.values())
        cost_basis = sum(h.avg_price * h.amount for h in self._holdings.values())
        pnl = total_value - cost_basis
        pnl_pct = (pnl / cost_basis * 100.0) if abs(cost_basis) > 1e-12 else 0.0
        return Portfolio(
            holdings=dict(self._holdings),
            total_value=total_value,
            pnl=pnl,
            pnl_pct=pnl_pct,
        )

    def update_holding(self, token: str, amount: float, price: float) -> None:
        if token not in self._holdings:
            self._holdings[token] = Holding(
                token=token,
                amount=0.0,
                avg_price=0.0,
                current_price=price,
                value=0.0,
                pnl=0.0,
            )

        h = self._holdings[token]

        if amount > 0:
            total_cost = h.avg_price * h.amount + price * amount
            h.amount += amount
            if h.amount > 1e-12:
                h.avg_price = total_cost / h.amount
        else:
            h.amount = max(0.0, h.amount + amount)

        h.current_price = price
        h.value = h.amount * h.current_price
        h.pnl = (h.current_price - h.avg_price) * h.amount

        if h.amount < 1e-12:
            del self._holdings[token]

    def update_price(self, token: str, price: float) -> None:
        if token in self._holdings:
            h = self._holdings[token]
            h.current_price = price
            h.value = h.amount * price
            h.pnl = (price - h.avg_price) * h.amount

    def record_trade(self, trade: Trade) -> None:
        if trade.status == "Executed":
            if trade.direction == "BUY":
                self.update_holding(trade.token, trade.amount, trade.price)
            else:
                self.update_holding(trade.token, -trade.amount, trade.price)
        self._trade_history.append(trade)

    def get_trade_history(self, limit: int = 0) -> list[Trade]:
        if limit <= 0 or limit >= len(self._trade_history):
            return list(self._trade_history)
        return list(self._trade_history[-limit:])

    def calculate_pnl(self) -> tuple[float, float]:
        portfolio = self.get_portfolio()
        pnl = portfolio.pnl
        pnl_pct = (
            (pnl / self._initial_value * 100.0)
            if abs(self._initial_value) > 1e-12
            else portfolio.pnl_pct
        )
        return pnl, pnl_pct

    @property
    def holdings_count(self) -> int:
        return len(self._holdings)

    @property
    def trade_count(self) -> int:
        return len(self._trade_history)

File: plugins/plugin-auto-trader/rust/src/portfolio.rs
use std::collections::HashMap;

use crate::types::{Holding, Portfolio, Trade, TradeDirection, TradeStatus};

/// Manages an in-memory portfolio: holdings and trade history.
pub struct PortfolioManager {
    holdings: HashMap<String, Holding>,
    trade_history: Vec<Trade>,
    initial_value: f64,
}

impl PortfolioManager {
    pub fn new(initial_value: f64) -> Self {
        Self {
            holdings: HashMap::new(),
            trade_history: Vec::new(),
            initial_value,
        }
    }

    /// Return a snapshot of the current portfolio.
    pub fn get_portfolio(&self) -> Portfolio {
        let total_value: f64 = self.holdings.values().map(|h| h.value).sum();
        let cost_basis: f64 = self
            .holdings
            .values()
            .map(|h| h.avg_price * h.amount)
            .sum();
        let pnl = total_value - cost_basis;
        let pnl_pct = if cost_basis.abs() > f64::EPSILON {
            pnl / cost_basis * 100.0
        } else {
            0.0
        };

        Portfolio {
            holdings: self.holdings.clone(),
            total_value,
            pnl,
            pnl_pct,
        }
    }

    /// Update (or create) a holding for the given token.
    pub fn update_holding(&mut self, token: &str, amount: f64, price: f64) {
        let holding = self.holdings.entry(token.to_string()).or_insert(Holding {
            token: token.to_string(),
            amount: 0.0,
            avg_price: 0.0,
            current_price: price,
            value: 0.0,
            pnl: 0.0,
        });

        if amount > 0.0 {
            // Weighted average price
            let total_cost = holding.avg_price * holding.amount + price * amount;
            holding.amount += amount;
            if holding.amount > f64::EPSILON {
                holding.avg_price = total_cost / holding.amount;
            }
        } else {
            // Reducing position
            holding.amount = (holding.amount + amount).max(0.0);
        }

        holding.current_price = price;
        holding.value = holding.amount * holding.current_price;
        holding.pnl = (holding.current_price - holding.avg_price) * holding.amount;

        // Remove holdings with zero amount
        if holding.amount < f64::EPSILON {
            self.holdings.remove(token);
        }
    }

    /// Update the current price for a token without changing the amount.
    pub fn update_price(&mut self, token: &str, price: f64) {
        if let Some(h) = self.holdings.get_mut(token) {
            h.current_price = price;
            h.value = h.amount * price;
            h.pnl = (price - h.avg_price) * h.amount;
        }
    }

    /// Record a completed trade.
    pub fn record_trade(&mut self, trade: Trade) {
        if trade.status == TradeStatus::Executed {
            match trade.direction {
                TradeDirection::Buy => {
                    self.update_holding(&trade.token, trade.amount, trade.price);
                }
                TradeDirection::Sell => {
                    self.update_holding(&trade.token, -trade.amount, trade.price);
                }
            }
        }
        self.trade_history.push(trade);
    }

    /// Return the most recent `limit` trades (or all if limit is 0).
    pub fn get_trade_history(&self, limit: usize) -> Vec<Trade> {
        if limit == 0 || limit >= self.trade_history.len() {
            return self.trade_history.clone();
        }
        self.trade_history[self.trade_history.len() - limit..].to_vec()
    }

    /// Calculate total PnL: (absolute, percentage).
    pub fn calculate_pnl(&self) -> (f64, f64) {
        let portfolio = self.get_portfolio();
        let pnl = portfolio.pnl;
        let pnl_pct = if self.initial_value.abs() > f64::EPSILON {
            pnl / self.initial_value * 100.0
        } else {
            portfolio.pnl_pct
        };
        (pnl, pnl_pct)
    }

    /// Number of holdings currently tracked.
    pub fn holdings_count(&self) -> usize {
        self.holdings.len()
    }

    /// Total number of trades recorded.
    pub fn trade_count(&self) -> usize {
        self.trade_history.len()
    }
}
